### PR TITLE
Import tests from Google V8 (templates)

### PR DESCRIPTION
**Pull request by jugglinmike**
These tests are derived from the following files within the Google V8
project:

    test/mjsunit/es6/templates.js

Some of these tests include non-printable characters, causing git to
infer that they are binary files. Introduce a `.gitattributes` file to
configure git to consistently display the source text of all JavaScript
files.

### CODE REVIEW

File: test/built-ins/String/raw/special-characters.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  'Unicode escape sequences'
+);
+assert.sameValue(
+  String.raw`\
\
+\
+`,
+  '\\\n\\\n\\\n',
+  'Literal characters'
+);
```
Comment (arv) [inline test/built-ins/String/raw/special-characters.js]: Missing `);`

Comment (arv) [inline test/built-ins/String/raw/special-characters.js]: nm, the diff was stupid

File: test/language/templates/expressions/function-invocation.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Copyright (C) Copyright 2014 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+es6id: 12.2.8
+description: >
+    Expressions should be evaluated and converted to Stings accordingto the
```
Comment (arv) [inline test/language/templates/expressions/function-invocation.js]: s/Stings/Strings/

Comment (arv) [inline test/language/templates/expressions/function-invocation.js]: s/accordingto/according to/

File: test/language/templates/expressions/legacy-octal-escape-sequence-non-strict.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+es6id: 12.2.8
+description: >
+    The expression within the template should be evaluated according to the
+    semantics of the surrounding context.
+    The SV of EscapeSequence :: HexEscapeSequence is the SV of the
+    HexEscapeSequence.
+flags: [noStrict]
+---*/
+
+assert.sameValue(`${'\07'}`, '');
```
Comment (arv) [inline test/language/templates/expressions/legacy-octal-escape-sequence-non-strict.js]: This is Annex B stuff? How do we handle that in general?

File: test/language/templates/expressions/method-invocation.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// Copyright (C) Copyright 2014 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+es6id: 12.2.8
+description: >
+    Expressions should be evaluated and converted to Stings accordingto the
```
Comment (arv) [inline test/language/templates/expressions/method-invocation.js]: same typo

File: test/language/templates/tagged/call-expression-argument-list-evaluation.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+calls = 0;
+(function(site, n, s, o, f, r) {
+  calls++;
+  assert.sameValue(arguments.length, 6);
+  assert.sameValue(n, number);
+  assert.sameValue(s, string);
+  assert.sameValue(o, object);
+  assert.sameValue(f, fn);
+  assert.sameValue(r, 'result');
+  assert.sameValue(arguments.length, 6, 'TemplateHead arguments length');
```
Comment (arv) [inline test/language/templates/tagged/call-expression-argument-list-evaluation.js]: repeat of line 30

File: test/language/templates/tagged/template-object.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    defined by the GetTemplateObject abstract operation.
+includes: [propertyHelper.js]
+---*/
+var templateObject;
+
+(function(parameter) {
+  templateObject = parameter;
+})`${1}`;
+
+templateObject.test262Prop = true;
```
Comment (arv) [inline test/language/templates/tagged/template-object.js]: this will throw in strict mode

Comment (jugglinmike) [inline test/language/templates/tagged/template-object.js]: Good point. To limit duplication, I've removed just the assertions for "frozen-ness" and created two new files for them: one runs only in strict mode and expects an error, the other runs only in non-strict mode and simply asserts that the assignment is not honored.

File: test/language/templates/tagged/call-expression-argument-list-evaluation.js (hunk 5 above)
Comment (jugglinmike) [inline test/language/templates/tagged/call-expression-argument-list-evaluation.js]: I removed line 30 since it lacks an assertion message

File: test/language/templates/expressions/method-invocation.js (hunk 4 above)
Comment (jugglinmike) [inline test/language/templates/expressions/method-invocation.js]: Grepping the changeset, this shows up 6 times. I've corrected all occurrences.

File: test/language/templates/expressions/function-invocation.js (hunk 2 above)
Comment (jugglinmike) [inline test/language/templates/expressions/function-invocation.js]: Grepping the changeset, this shows up 6 times. I've corrected all occurrences.

File: test/built-ins/String/raw/special-characters.js (hunk 1 above)
Comment (jugglinmike) [inline test/built-ins/String/raw/special-characters.js]: Yeah, I've run into a few instances where non-printable characters break GitHub's UI while writing these tests!

File: test/language/templates/expressions/legacy-octal-escape-sequence-non-strict.js (hunk 3 above)
Comment (jugglinmike) [inline test/language/templates/expressions/legacy-octal-escape-sequence-non-strict.js]: I don't know that. Do you have any thoughts on this, @bterlson?

Comment (caitp) [inline test/language/templates/expressions/legacy-octal-escape-sequence-non-strict.js]: annex b stuff belongs in test/annexB, I think :> --- no reason template-related stuff couldn't go in there?

Comment (jugglinmike) [inline test/language/templates/expressions/legacy-octal-escape-sequence-non-strict.js]: Definitely. I should have known this because I've already added some tests there (for some reason, I was thinking it was Annex E...)

**Comment by anba** [comment]
`language/templates/template-values/non-escape-character.js` should be removed for now because it's not covered by the spec (see also https://bugs.ecmascript.org/show_bug.cgi?id=3477).

**Comment by bterlson** [comment]
templates/expressions --> expressions/template
templates/tagged --> expressions/template (or expressions/tagged-template, although I think it's fine to club them together)

File: test/language/templates/template-values/line-continuation.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+var calls;
+
+calls = 0;
+(function(cs) {
+  calls++;
+  assert.sameValue(cs.raw[0], '\u005C\n\u005C\n\u005C\n');
+})`\
+\
+\
`
```
Comment (bterlson) [inline test/language/templates/template-values/line-continuation.js]: I think this template literal should also test <LS>, <PS>, and <CR><LF> (similar to line-terminator-sequence?)

File: test/language/templates/template-values/line-terminator-sequence.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// Copyright (C) Copyright 2014 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+es6id: 11.8.6
+description: >
+    The TRV of LineTerminatorSequence :: <LF> is the code unit value 0x000A.
+    The TRV of LineTerminatorSequence :: <CR> is the code unit value 0x000A.
+    The TRV of LineTerminatorSequence :: <CR><LF> is the sequence consisting of
+    the code unit value 0x000A.
```
Comment (bterlson) [inline test/language/templates/template-values/line-terminator-sequence.js]: Add <LS> and <PS>?

Comment (jugglinmike) [inline test/language/templates/template-values/line-terminator-sequence.js]: Sorry, I don't understand. Can you rephrase?

Comment (bterlson) [inline test/language/templates/template-values/line-terminator-sequence.js]: See the antepenultimate and penultimate bullets of [11.8.6.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-static-semantics-tv-s-and-trv-s)

Comment (jugglinmike) [inline test/language/templates/template-values/line-terminator-sequence.js]: Ah, now I understand. (I think GitHub was initially having some trouble with the `<` and `>` characters in your comment.) This batch of tests is ported from V8 directly, but I can certainly include more assertions. I'll just introduce them with a standalone commit to avoid confusion.

**Comment by jugglinmike** [comment]
I've addressed all the feedback I received. Thanks for the review!

File: test/annexB/templates/legacy-octal-escape-sequence-non-strict.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+es6id: 12.2.8
+description: >
+    The expression within the template should be evaluated according to the
+    semantics of the surrounding context.
+    The SV of EscapeSequence :: HexEscapeSequence is the SV of the
+    HexEscapeSequence.
+flags: [noStrict]
+---*/
+
+assert.sameValue(`${'\07'}`, '');
```
Comment (caitp) [inline test/annexB/templates/legacy-octal-escape-sequence-non-strict.js]: I'm still kind of keen on using hex escapes or unicode escapes for the expected values here --- I know it means using another language feature to test this one, but it's a lot easier to see what's going on on github, and in many text editors, too.

**Comment by bterlson** [comment]
Merged without squash. Oh well :-P

File: test/annexB/templates/legacy-octal-escape-sequence-non-strict.js (hunk 9 above)
Comment (bterlson) [inline test/annexB/templates/legacy-octal-escape-sequence-non-strict.js]: These tests are really hard to debug. I would be ok with using escapes for expected values.
